include/gatekeeper_gk.h

```diff
@@ -118,10 +151,24 @@ struct gk_fib {
 
 			/*
 			 * Cache the whole Ethernet header when the @next_fib
-			 * action is GK_FWD_NEIGHBOR.
+			 * action is GK_FWD_NEIGHBOR_*_NET.
 			 */
-			struct ether_cache eth_cache;
+			struct ether_cache *eth_cache;
 		} grantor;
+
+		/*
+		 * When the action is GK_FWD_NEIGHBOR_*_NET, it stores all
+		 * the neighbors' Ethernet headers in a hash table.
+		 * The entries can be accessed according to its IP address.
+		 */
+		struct {
+			/*
+			 * Since IPv4 entries can be accessed easily, we store
+			 * the IPv4 and IPv6 entries separately.
+			 */
+			struct neighbor_hash_table neigh_tbl;
+			struct neighbor_hash_table neigh_tbl6;
```

Comment:
If I got the code right, there should be a `struct neigh` and another `struct neigh6`, each only holding its hash table of neighbors. The reason for this is that is that a FIB entry in the IPv4 LPM will only return IPv4 neighbors. The same goes for an IPv6 FIB.